node/src/accountant/scanners.rs

```diff
@@ -0,0 +1,1871 @@
+
+    fn mark_as_started(&mut self, timestamp: SystemTime) {
+        self.common.initiated_at_opt = Some(timestamp);
+    }
+
+    fn mark_as_ended(&mut self, logger: &Logger) {
+        match self.scan_started_at() {
+            Some(timestamp) => {
+                let elapsed_time = SystemTime::now()
+                    .duration_since(timestamp)
+                    .expect("Unable to calculate elapsed time for the scan.")
+                    .as_millis();
+                info!(logger, "The Receivable scan ended in {elapsed_time}ms.");
+                self.common.initiated_at_opt = None;
+            }
+            None => error!(
+                logger,
+                "The scan_finished() was called for Receivable scanner but timestamp was not found"
+            ),
+        };
+    }
+
+    as_any_impl!();
+}
+
+impl ReceivableScanner {
+    pub fn new(
+        receivable_dao: Box<dyn ReceivableDao>,
+        banned_dao: Box<dyn BannedDao>,
+        payment_thresholds: Rc<PaymentThresholds>,
+        earning_wallet: Rc<Wallet>,
+        financial_statistics: Rc<RefCell<FinancialStatistics>>,
+    ) -> Self {
+        Self {
+            common: ScannerCommon::new(payment_thresholds),
+            earning_wallet,
+            receivable_dao,
+            banned_dao,
+            financial_statistics,
+        }
+    }
+
+    pub fn scan_for_delinquencies(&self, timestamp: SystemTime, logger: &Logger) {
+        info!(logger, "Scanning for delinquencies");
+        self.ban_nodes(timestamp, logger);
+        self.unban_nodes(timestamp, logger);
+    }
+
+    pub fn ban_nodes(&self, timestamp: SystemTime, logger: &Logger) {
+        self.receivable_dao
+            .new_delinquencies(timestamp, self.common.payment_thresholds.as_ref())
+            .into_iter()
+            .for_each(|account| {
+                self.banned_dao.ban(&account.wallet);
+                let (balance, age) = balance_and_age(timestamp, &account);
+                info!(
+                    logger,
+                    "Wallet {} (balance: {} MASQ, age: {} sec) banned for delinquency",
+                    account.wallet,
+                    balance,
+                    age.as_secs()
+                )
+            });
+    }
+
+    pub fn unban_nodes(&self, timestamp: SystemTime, logger: &Logger) {
+        self.receivable_dao
+            .paid_delinquencies(self.common.payment_thresholds.as_ref())
+            .into_iter()
+            .for_each(|account| {
+                self.banned_dao.unban(&account.wallet);
+                let (balance, age) = balance_and_age(timestamp, &account);
+                info!(
+                    logger,
+                    "Wallet {} (balance: {} MASQ, age: {} sec) is no longer delinquent: unbanned",
+                    account.wallet,
+                    balance,
+                    age.as_secs()
+                )
+            });
+    }
+
+    pub fn financial_statistics(&self) -> FinancialStatistics {
```

Comment:
Once you use my suggested changes with the simplified assignments (+=), you will not need this anymore. There is a test which uses it too but it can work easily without such a fancy method.